model/stream.go

```diff
@@ -236,3 +241,157 @@ func (s Stream) Attachments() []File {
+	bufPtr := 0
+
+	for bufPtr < len(buf) && visible < maxlen {
+
+		// Move to nearest tag and count visible characters along the way.
+		offset := 0
+		visibleCharacterMaxReached := false
+		entityDetected := false
+
+		for localOffset, runeValue := range string(buf[bufPtr:]) {
+			offset = localOffset
+
+			if runeValue == '<' {
+				// Start of tag.
+				break
+			} else if runeValue == '&' {
+				// Possible start of HTML Entity
+				loc := EntityExpr.FindIndex(buf[bufPtr+localOffset:])
+				if loc != nil && loc[0] == 0 {
+					// Entity found!
+					entityDetected = true
+					offset += loc[1] - 1 // Now pointing to ;
+				}
+				visible += 1
+			} else if unicode.IsPrint(runeValue) && !unicode.IsSpace(runeValue) {
+				// Printable, non-space character. Increment visible count.
+				visible += 1
+			}
+
+			// Check if the limit of visible characters has been reached.
+			if visible >= maxlen {
+				visibleCharacterMaxReached = true
+				break
+			}
+
+			if entityDetected {
+				break
+			}
+		}
+
+		// Increment bufPtr to end of scanned section
+		bufPtr += offset
+
+		// Stop scanning if the end of the buffer was reached or if the max
+		// desired visible characters was reached
+		if visibleCharacterMaxReached || bufPtr >= len(buf)-1 {
+			break
+		}
+
+		// If an entity was detected, continue scanning for next tag
+		if entityDetected {
+			// Advance past the ;
+			bufPtr += 1
+			continue
+		}
+
+		// Now find the expression sub-matches
+		matches := TagExpr.FindSubmatch(buf[bufPtr:])
+		tagName := string(matches[2])
+
+		// Advance pointer to the end of the tag
+		bufPtr += len(matches[0])
+
+		// If this is a void element, do not count it as a start tag
+		isVoidElement := false
+		for _, voidElementTagName := range voidElementTags {
+			if tagName == voidElementTagName {
+				isVoidElement = true
+				break
+			}
+		}
+		if isVoidElement {
+			continue
+		}
+
+		isStartTag := len(matches[1]) == 0
+
+		if isStartTag {
+			// This is a start tag. Push the tag to the stack.
+			tagStack = append(tagStack, tagName)
+		} else {
+			// This is an end tag. First, check to make sure the end tag is
+			// matches what's on top of the stack.
+			if len(tagStack) == 0 || tagStack[len(tagStack)-1] != tagName {
+				return nil, ErrUnbalancedTags
+			}
+
+			// Now, pop the tag stack.
+			tagStack = tagStack[0 : len(tagStack)-1]
+		}
+	}
+
+	// At this point, bufPtr points to the last rune that should be copied to
+	// the output stream. Increment bufPtr past this rune, turning bufPtr into
+	// the number of bytes that should be copied.
+	_, size := utf8.DecodeRune(buf[bufPtr:])
+	bufPtr += size
+
+	// Copy the desired input to the output buffer.
+	output := buf[0:bufPtr]
+
+	// Copy ellipsis
+	output = append(output, []byte(ellipsis)...)
+
+	// Finally, create a closing tag for each tag in the stack.
+	for i := len(tagStack) - 1; i >= 0; i-- {
+		output = append(output, []byte(fmt.Sprintf("</%s>", tagStack[i]))...)
+	}
+
+	return output, nil
+}
```

Comment:
Let's put all this in the `tools` package, having it in `model` feels off

Reply:
Agree, I also had to take a bit of a different approach now, as using tools would have introduced a circular dependency, that would be very ugly to solve. It is kinda of a problem that `tools` depends on `dao` and `model`. Maybe we can decouple it somehow in the future.